src/java/org/apache/cassandra/index/sai/utils/PrimaryKey.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.ClusteringComparator;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+/**
+ * Representation of the primary key for a row consisting of the {@link DecoratedKey} and
+ * {@link Clustering} associated with a {@link org.apache.cassandra.db.rows.Row}.
+ */
+public class PrimaryKey implements Comparable<PrimaryKey>
+{
+    private final Token token;
+    private final DecoratedKey partitionKey;
+    private final Clustering<?> clustering;
+    private final ClusteringComparator clusteringComparator;
+
+    public PrimaryKey(Token token,
```

Comment:
nit: Reduce visibility to default/package-private perhaps, given the static factories? That makes it harder for someone to create an invalid combination, like a non-null clustering w/ a null partition key, but it still might be a good idea to have a preconditions check to make sure we've got a valid one. (I think we could also just pass a null comparator too, if we make one other change here: https://github.com/maedhroz/cassandra/pull/6/files#r1048997860)

Reply:
I have done this and added a separate constructor for `Token` only `PrimaryKey`s. I tried adding preconditions to the main constructor but they seemed to add more complexity and didn't make the code any easier to read. I think that by only accessing the constructors through specific creators in the factory is enough to ensure correctness.